### DISCUSSION TITLE
CompositionOfAir class & LifeSupport Function

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#538] [General]
@mokun I've noticed that the CompositionOfAir class is actually implementing a capability of the Building class. The Building should be responsible for it's own composition stats. Plus doesn't the calculation overlap with the role of the LifeSupport function?
If a Building has no LifeSupport function then it has no air composition. Only inhabitable Buildings can have an Air Composition I think.
So I think we may have some expensive overlapping logic.

**Comment by mokun** [comment]
>  CompositionOfAir class is actually implementing a capability of the Building class. 

For now, CompositionOfAir would perform all the atmospheric calculation for each habitable buildings of a settlement.

Yes perhaps we can have each building calculate its own atmosphere instead.

**Reply by mokun** [reply]
> expensive overlapping logic.

Let's use this class inside an habitable building. This way, it will no longer need to generate a new list of habitable buildings on each frame :

```java
List<Building> newList = unitManager.getSettlementByID(settlementID).getBuildingManager().getBuildingsWithLifeSupport();

for (Building b : newList) {
	int id = b.getInhabitableID();
	....
}
...
```

**Reply by bevans2000** [reply]
Sounds good.  I've created an Issue from this. #543
